File: ruby/collatz-conjecture/collatz_conjecture.rb
module CollatzConjecture
  def self.steps(number)
    raise ArgumentError, "Number must be > 0" unless number > 0

    count = 0

    until number == 1 do
      number = number.even? ? number / 2 : number * 3 + 1
      count += 1
    end

    count
  end
end

File: clojure/collatz-conjecture/src/collatz_conjecture.clj
(ns collatz-conjecture)

(defn- step [n] (if (even? n) (quot n 2) (inc (* n 3))))
(defn- collatz-sequence [n] (take-while #(not= 1 %) (iterate step n)))

(defn collatz [n]
  (if (< n 1)
    (throw (IllegalArgumentException. "Number must be greater than or equal to 1"))
    (count (collatz-sequence n))))
